### ISSUE TITLE
Seeking Help with Development of a Program for Monitoring, Prediction, and Assistance in Cannabis Cultivation

### ISSUE THREAD

**Issue by shegapt** [shegapt/IAC#1]
Hello everyone,
I hope you are doing well. I am currently working on developing a program for monitoring, prediction, and assistance in cannabis cultivation, and I am facing some challenges that I would like to discuss and resolve with the help of the community.

Project Description:
My project involves creating an automated system for monitoring, predicting, and guiding cannabis cultivation. The program includes a chatbot that is intended to provide detailed guidance on different cannabis strains, optimal environmental conditions, cultivation techniques, and plant maintenance.

Technologies Used:
Language: Python
Key Libraries: ChatterBot, Tkinter, SQLite (or another recommended database)

Issue Faced:
When attempting to implement ChatterBot with a suitable storage adapter, I am encountering compatibility issues with the current version of SQLAlchemy and Python. The program is throwing an error related to the time.clock() function, which has been deprecated in recent Python versions.

Request for Assistance:
If anyone can provide guidance on resolving the issue with SQLAlchemy and ChatterBot or suggest an alternative approach to implementing an effective chatbot for my project, I would greatly appreciate it.

Thank you in advance for your assistance, and I look forward to learning from the community's experience and knowledge.
Best regards, ShegaPT

CODE:
import tkinter as tk
from tkinter import ttk, scrolledtext, messagebox
from tkcalendar import DateEntry
from datetime import datetime
import json
import os
from chatterbot import ChatBot
from chatterbot.trainers import ListTrainer

# Data files
MEMORY_FILE = 'cultivation_data.json'
USER_DATA_FILE = "user_data.json"

# Initial memory data
memory = {
    'data': [],
    'plantas': [],
    'ambientes': [],
    'calculos': [],
}

user_memory = {
    "plantas": [],
    "ambientes": [],
    "calculos": [],
    "dados_diarios": [],
}

location_data = {
    "cidade": "",
    "país": "",
}

# Load data from JSON file
def load_data(file):
    if os.path.exists(file):
        with open(file, "r") as f:
            return json.load(f)
    else:
        return None

# Save data to JSON file
def save_data(data, file):
    with open(file, "w") as f:
        json.dump(data, f, indent=4)

# Load initial memory
loaded_memory = load_data(MEMORY_FILE)
if loaded_memory:
    memory = loaded_memory
else:
    save_data(memory, MEMORY_FILE)

# Initialize Chatbot
bot = ChatBot(
    'Bot',
    storage_adapter='chatterbot.storage.SQLStorageAdapter',
    database_uri='sqlite:///database.sqlite3'
)

trainer = ListTrainer(bot)

# Initial Chatbot training
trainer.train([
    'Hello',
    'Hello, how can I help you?',
    'How are you?',
    'I'm fine, thank you. And you?',
    'I'm great!',
    'I need tips for cannabis cultivation',
    'Sure, do you have any specific strain in mind?',
    'Yes, I'm interested in an Indica strain',
    'Great! What is the temperature and humidity of your environment?',
    'The temperature is around 25°C and humidity is 60%',
    'Those conditions are good for Indica. Make sure to maintain good ventilation.',
    'Thanks for the tips.',
    'You're welcome! Remember to monitor your plant regularly.'
])

# Load user data from JSON file
def load_user_data(file):
    if os.path.exists(file):
        with open(file, "r") as f:
            return json.load(f)
    else:
        return None

# Save user data to JSON file
def save_user_data(data, file):
    with open(file, "w") as f:
        json.dump(data, f, indent=4)

# Load initial user memory
loaded_user_data = load_user_data(USER_DATA_FILE)
if loaded_user_data:
    user_memory = loaded_user_data
else:
    save_user_data(user_memory, USER_DATA_FILE)

# Function to collect daily data
def collect_data(root, selected_plant):
    collection_window = tk.Toplevel(root)
    collection_window.title("Add Daily Data")
    collection_window.geometry("600x500")

    tk.Label(collection_window, text="Date (DD-MM-YYYY):").grid(row=0, column=0, padx=10, pady=10)
    date_entry = DateEntry(collection_window, width=12, background='darkblue', foreground='white', borderwidth=2)
    date_entry.grid(row=0, column=1, padx=10, pady=10)

    tk.Label(collection_window, text="Time (HH:MM):").grid(row=1, column=0, padx=10, pady=10)
    time_entry = tk.Entry(collection_window)
    time_entry.grid(row=1, column=1, padx=10, pady=10)

    tk.Label(collection_window, text="Water Addition:").grid(row=2, column=0, padx=10, pady=10)
    water_check = tk.Checkbutton(collection_window)
    water_check.grid(row=2, column=1, padx=10, pady=10)

    tk.Label(collection_window, text="Nutrient Addition:").grid(row=3, column=0, padx=10, pady=10)
    nutrient_check = tk.Checkbutton(collection_window)
    nutrient_check.grid(row=3, column=1, padx=10, pady=10)

    tk.Label(collection_window, text="Insecticide Addition:").grid(row=4, column=0, padx=10, pady=10)
    insecticide_entry = tk.Entry(collection_window)
    insecticide_entry.grid(row=4, column=1, padx=10, pady=10)

    tk.Label(collection_window, text="Plant Size:").grid(row=5, column=0, padx=10, pady=10)
    size_entry = tk.Entry(collection_window)
    size_entry.grid(row=5, column=1, padx=10, pady=10)

    tk.Label(collection_window, text="PPM:").grid(row=6, column=0, padx=10, pady=10)
    ppm_entry = tk.Entry(collection_window)
    ppm_entry.grid(row=6, column=1, padx=10, pady=10)

    tk.Label(collection_window, text="EC:").grid(row=7, column=0, padx=10, pady=10)
    ec_entry = tk.Entry(collection_window)
    ec_entry.grid(row=7, column=1, padx=10, pady=10)

    tk.Label(collection_window, text="pH:").grid(row=8, column=0, padx=10, pady=10)
    ph_entry = tk.Entry(collection_window)
    ph_entry.grid(row=8, column=1, padx=10, pady=10)

    def save_daily_data():
        data = date_entry.get_date().strftime('%d-%m-%Y') if date_entry.get_date() else ""
        time = time_entry.get()
        water = water_check.instate(['selected'])
        nutrient = nutrient_check.instate(['selected'])
        insecticide = insecticide_entry.get()
        size = size_entry.get()
        ppm = ppm_entry.get()
        ec = ec_entry.get()
        ph = ph_entry.get()

        if not data or not time:
            messagebox.showerror("Error", "Please fill in both date and time.")
            return

        if not water and not nutrient:
            messagebox.showerror("Error", "Select at least Water Addition or Nutrient Addition.")
            return

        daily_entry = {
            "data": data,
            "hora": time,
            "agua": water,
            "nutrientes": nutrient,
            "inseticida": insecticide,
            "tamanho": size,
            "ppm": ppm,
            "ec": ec,
            "ph": ph,
        }

        for plant in user_memory["plantas"]:
            if plant["nome"] == selected_plant:
                plant["dados_diarios"].append(daily_entry)
                break

        save_user_data(user_memory, USER_DATA_FILE)
        messagebox.showinfo("Saved", "Daily data saved successfully!")
        collection_window.destroy()

    tk.Button(collection_window, text="Save", command=save_daily_data).grid(row=9, column=0, columnspan=2, pady=20)

# Function to view daily data
def view_daily_data(root, selected_plant):
    view_window = tk.Toplevel(root)
    view_window.title(f"Daily Data for {selected_plant}")
    view_window.geometry("800x600")

    # Get data for selected plant
    plant_data = None
    for plant in user_memory["plantas"]:
        if plant["nome"] == selected_plant:
            plant_data = plant
            break

    if not plant_data or "dados_diarios" not in plant_data or not plant_data["dados_diarios"]:
        tk.Label(view_window, text="No daily data available for this plant.").pack(pady=20)
        return

    # Create table to display daily data
    tree = ttk.Treeview(view_window)
    tree["columns"] = ("Data", "Hora", "Water", "Nutrients", "Insecticide", "Size", "PPM", "EC", "pH")
    tree.column("#0", width=0, stretch=tk.NO)
    tree.column("Data", anchor=tk.CENTER, width=100)
    tree.column("Hora", anchor=tk.CENTER, width=100)
    tree.column("Water", anchor=tk.CENTER, width=80)
    tree.column("Nutrients", anchor=tk.CENTER, width=100)
    tree.column("Insecticide", anchor=tk.CENTER, width=100)
    tree.column("Size", anchor=tk.CENTER, width=80)
    tree.column("PPM", anchor=tk.CENTER, width=80)
    tree.column("EC", anchor=tk.CENTER, width=80)
    tree.column("pH", anchor=tk.CENTER, width=80)

    tree.heading("#0", text="", anchor=tk.CENTER)
    tree.heading("Data", text="Data", anchor=tk.CENTER)
    tree.heading("Hora", text="Hora", anchor=tk.CENTER)
    tree.heading("Water", text="Water", anchor=tk.CENTER)
    tree.heading("Nutrients", text="Nutrients", anchor=tk.CENTER)
    tree.heading("Insecticide", text="Insecticide", anchor=tk.CENTER)
    tree.heading("Size", text="Size", anchor=tk.CENTER)
    tree.heading("PPM", text="PPM", anchor=tk.CENTER)
    tree.heading("EC", text="EC", anchor=tk.CENTER)
    tree.heading("pH", text="pH", anchor=tk.CENTER)

    for data in plant_data["dados_diarios"]:
        tree.insert("", tk.END, values=(
            data["data"],
            data["hora"],
            "Yes" if data["agua"] else "No",
            "Yes" if data["nutrientes"] else "No",
            data["inseticida"],
            data["tamanho"],
            data["ppm"],
            data["ec"],
            data["ph"]
        ))

    tree.pack(pady=20, padx=20, fill=tk.BOTH, expand=True)

# Main function
def main():
    root = tk.Tk()
    root.title("Cannabis Cultivation Assistant")
    root.geometry("800x600")

    tk.Label(root, text="Welcome to the Cannabis Cultivation Assistant", font=("Helvetica", 16)).pack(pady=20)

    # Function to start a new cultivation
    def start_new_cultivation():
        plant_name = entry_plant_name.get()

        if not plant_name:
            messagebox.showerror("Error", "Please enter the plant name.")
            return

        for plant in user_memory["plantas"]:
            if plant["nome"].lower() == plant_name.lower():
                messagebox.showerror("Error", f"The plant '{plant_name}' is already in your list.")
                return

        user_memory["plantas"].append({
            "nome": plant_name,
            "dados_diarios": []
        })

        save_user_data(user_memory, USER_DATA_FILE)
        messagebox.showinfo("Success", f"Cultivation of plant '{plant_name}' started successfully.")
        new_cultivation_window.destroy()

    # Interface to start a new cultivation
    def new_cultivation_interface():
        global new_cultivation_window
        new_cultivation_window = tk.Toplevel(root)
        new_cultivation_window.title("New Cultivation")
        new_cultivation_window.geometry("400x200")

        tk.Label(new_cultivation_window, text="Plant Name:").grid(row=0, column=0, padx=10, pady=10)
        global entry_plant_name
        entry_plant_name = tk.Entry(new_cultivation_window)
        entry_plant_name.grid(row=0, column=1, padx=10, pady=10)

        tk.Button(new_cultivation_window, text="Start Cultivation", command=start_new_cultivation).grid(row=1, column=0, columnspan=2, pady=20)

    # Function to select a plant and view its daily data
    def select_plant():
        selected_plant = combo_plants.get()

        if not selected_plant:
            messagebox.showerror("Error", "Please select a plant.")
            return

        view_daily_data(root, selected_plant)

    # Load user plants
    user_plants = [plant["nome"] for plant in user_memory["plantas"]]

    # Main interface
    tk.Label(root, text="Select a Plant:", font=("Helvetica", 14)).pack(pady=10)

    global combo_plants
    combo_plants = ttk.Combobox(root, values=user_plants)
    combo_plants.pack(pady=10)

    tk.Button(root, text="Add New Plant", command=new_cultivation_interface).pack(pady=10)
    tk.Button(root, text="View Daily Data", command=select_plant).pack(pady=10)

    root.mainloop()

if __name__ == "__main__":
    main()
